File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/ApiRequestTyped.java
package spectable.common;

import java.util.ArrayList;
import java.util.List;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class ApiRequestTyped {
    public String method;
    public String page;
    public String parameter;
    public String body;

    public ApiRequestTyped(String method, String page, String parameter, String body) {
        this.method = method;
        this.page = page;
        this.parameter = parameter;
        this.body = body;
    }

    public ApiRequestTyped(ApiRequestString s) {
        this.method = s.method;
        this.page = s.page;
        this.parameter = s.parameter;
        this.body = s.body;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ApiRequestTyped)) return false;
        ApiRequestTyped that = (ApiRequestTyped) o;
        return Objects.equals(method, that.method)
            && Objects.equals(page, that.page)
            && Objects.equals(parameter, that.parameter)
            && Objects.equals(body, that.body);
    }

    @Override
    public int hashCode() {
        return Objects.hash(method, page, parameter, body);
    }

    public ApiRequestString toApiRequestString() {
        return new ApiRequestString(String.valueOf(method), String.valueOf(page), String.valueOf(parameter), String.valueOf(body));
    }

    public static List<ApiRequestTyped> fromStringList(List<ApiRequestString> list) {
        List<ApiRequestTyped> result = new ArrayList<>();
        for (ApiRequestString s : list) result.add(new ApiRequestTyped(s));
        return result;
    }

    public static List<ApiRequestString> toStringList(List<ApiRequestTyped> list) {
        List<ApiRequestString> result = new ArrayList<>();
        for (ApiRequestTyped t : list) result.add(t.toApiRequestString());
        return result;
    }

    /** Internal plumbing for {@link Json}; use toJSON() for JSON text. */
    public Map<String, Object> toJsonValue() {
        Map<String, Object> m = new LinkedHashMap<String, Object>();
        m.put("method", method);
        m.put("page", page);
        m.put("parameter", parameter);
        m.put("body", body);
        return m;
    }

    public String toJSON() { return Json.write(toJsonValue()); }

    /** Internal plumbing for {@link Json}; use fromJSON(String) for JSON text. */
    public static ApiRequestTyped fromJsonValue(Map<String, Object> m) {
        if (m == null) return null;
        return new ApiRequestTyped(
            Json.asString(Json.require(m, "method"), "method"),
            Json.asString(Json.require(m, "page"), "page"),
            Json.asString(Json.require(m, "parameter"), "parameter"),
            Json.asString(Json.require(m, "body"), "body"));
    }

    public static ApiRequestTyped fromJSON(String json) {
        return fromJsonValue(Json.parseObject(json));
    }

    public static String toJSONList(List<ApiRequestTyped> list) {
        List<Object> arr = new ArrayList<Object>();
        if (list != null)
            for (ApiRequestTyped item : list) arr.add(item == null ? null : item.toJsonValue());
        return Json.write(arr);
    }

    public static List<ApiRequestTyped> fromJSONList(String json) {
        List<ApiRequestTyped> result = new ArrayList<ApiRequestTyped>();
        for (Object e : Json.parseArray(json))
            result.add(fromJsonValue(Json.asObject(e, "ApiRequestTyped")));
        return result;
    }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/ApiRequestTyped.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using System;
using System.Buffers;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.Json;
using production;

    public class ApiRequestTyped
    {
        public string method;
        public string page;
        public string parameter;
        public string body;

        public ApiRequestTyped(string method, string page, string parameter, string body)
        {
            this.method = method;
            this.page = page;
            this.parameter = parameter;
            this.body = body;
        }

        public ApiRequestString ToApiRequestString()
        {
            return new ApiRequestString(
                Json.ToText(this.method),
                Json.ToText(this.page),
                Json.ToText(this.parameter),
                Json.ToText(this.body)
            );
        }

        public static List<ApiRequestString> ToStringList(List<ApiRequestTyped> list)
        {
            var result = new List<ApiRequestString>();
            foreach (var t in list) result.Add(t.ToApiRequestString());
            return result;
        }

        public static List<ApiRequestTyped> FromStringList(List<ApiRequestString> list)
        {
            var result = new List<ApiRequestTyped>();
            foreach (var s in list) result.Add(s.ToApiRequestTyped());
            return result;
        }

        public void WriteJson(Utf8JsonWriter w)
        {
            w.WriteStartObject();
            w.WriteString("method", this.method);
            w.WriteString("page", this.page);
            w.WriteString("parameter", this.parameter);
            w.WriteString("body", this.body);
            w.WriteEndObject();
        }

        public string ToJSON()
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer)) { WriteJson(w); }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static ApiRequestTyped FromJsonElement(JsonElement m)
        {
            return new ApiRequestTyped(
                Json.AsString(Json.Require(m, "method"), "method"),
                Json.AsString(Json.Require(m, "page"), "page"),
                Json.AsString(Json.Require(m, "parameter"), "parameter"),
                Json.AsString(Json.Require(m, "body"), "body")
            );
        }

        public static ApiRequestTyped FromJSON(string json)
        {
            return FromJsonElement(Json.Parse(json));
        }

        public static string ToJSONList(List<ApiRequestTyped> list)
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer))
            {
                w.WriteStartArray();
                foreach (var item in list) item.WriteJson(w);
                w.WriteEndArray();
            }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static List<ApiRequestTyped> FromJSONList(string json)
        {
            var result = new List<ApiRequestTyped>();
            var root = Json.Parse(json);
            Json.RequireArray(root, "ApiRequestTyped");
            foreach (var e in root.EnumerateArray()) result.Add(FromJsonElement(e));
            return result;
        }

        public override string ToString()
        {
            return $"Method={method}, Page={page}, Parameter={parameter}, Body={body}";
        }

        public override bool Equals(object? obj)
        {
            if (obj is not ApiRequestTyped other) return false;
            return object.Equals(this.method, other.method)
                && object.Equals(this.page, other.page)
                && object.Equals(this.parameter, other.parameter)
                && object.Equals(this.body, other.body);
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.method);
            h.Add(this.page);
            h.Add(this.parameter);
            h.Add(this.body);
            return h.ToHashCode();
        }
    }
}
